opentelemetry-sdk/src/opentelemetry/sdk/_metrics/_internal/metric_reader_storage.py

```diff
@@ -51,30 +57,61 @@ def _get_or_init_view_instrument_match(
         # Optimistically get the relevant views for the given instrument. Once set for a given
         # instrument, the mapping will never change
 
-        if instrument in self._view_instrument_match:
-            return self._view_instrument_match[instrument]
+        if instrument in self._instrument_view_instrument_matches:
+            return self._instrument_view_instrument_matches[instrument]
 
         with self._lock:
             # double check if it was set before we held the lock
-            if instrument in self._view_instrument_match:
-                return self._view_instrument_match[instrument]
+            if instrument in self._instrument_view_instrument_matches:
+                return self._instrument_view_instrument_matches[instrument]
 
             # not present, hold the lock and add a new mapping
             view_instrument_matches = []
             for view in self._sdk_config.views:
                 # pylint: disable=protected-access
                 if view._match(instrument):
-                    view_instrument_matches.append(
-                        _ViewInstrumentMatch(
-                            view=view,
-                            instrument=instrument,
-                            sdk_config=self._sdk_config,
-                            instrument_class_aggregation=(
-                                self._instrument_class_aggregation
-                            ),
-                        )
+
+                    new_view_instrument_match = _ViewInstrumentMatch(
+                        view=view,
+                        instrument=instrument,
+                        sdk_config=self._sdk_config,
+                        instrument_class_aggregation=(
+                            self._instrument_class_aggregation
+                        ),
                     )
 
+                    if isinstance(instrument, Asynchronous) and isinstance(
+                        view._aggregation, ExplicitBucketHistogramAggregation
+                    ):
+                        _logger.warning(
+                            "View %s and instrument %s will produce "
+                            "semantic errors when matched, the view "
+                            "has not been applied.",
+                            view,
+                            instrument,
+                        )
+                        continue
+
+                    for (
```

Comment:
Weren't we going to keep a dictionary/set of `ViewInstrumentMatches`? That way we don't have to have two for loops to check for existence?

Reply:
I see the point but after taking a look at the code I am not sure if it is possible or convenient to do this. Right now we have all `_ViewInstrumentMatches` grouped in separate lists. Each one of these lists is associated with an instrument. So, to check all `_ViewInstrumentMatches` we would need a separate `set` that would contain duplicate entries of all `_ViewInstrumentMatches`. On top of that, we would need to implement a `__hash__` function for `_ViewInstrumentMatch` that would serve the purpose of the current `conflicts` function (and we just agreed on using `conflicts` since it is much more semantically accurate). So, even when I recognize there may be a way to avoid looping through all the `_ViewInstrumentMatches`, I don't see the performance improvements that valuable for the reasons before. In any case, this is something we can address in a separate issue/PR :slightly_smiling_face:

Reply:
Yeah not a blocker.